plugins/woocommerce/src/Caching/ObjectCache.php

```diff
@@ -0,0 +1,405 @@
+			throw new CacheException( "Object id must be an int, a string, or null for 'set'", $this, $id );
+		}
+
+		$this->verify_expiration_value( $expiration );
+
+		if ( null === $id ) {
+			$id = $this->get_object_id( $object );
+			if ( null === $id ) {
+				throw new CacheException( "Null id supplied and the cache class doesn't implement get_object_id", $this );
+			}
+		}
+
+		$errors = $this->validate( $object );
+		if ( null !== $errors && 1 === count( $errors ) ) {
+			throw new CacheException( 'Object validation/serialization failed: ' . $errors[0], $this, $id, $errors );
+		} elseif ( ! empty( $errors ) ) {
+			throw new CacheException( 'Object validation/serialization failed', $this, $id, $errors );
+		}
+
+		$data = $this->serialize( $object );
+
+		/**
+		 * Filters the serialized object that will be passed by an instance of ObjectCache to its cache engine to be cached.
+		 *
+		 * @since 6.8.0
+		 *
+		 * @param array $data The already serialized object data.
+		 * @param array|object $object The object before serialization.
+		 * @returns array The actual serialized object data that will be passed to the cache engine.
+		 */
+		$data = apply_filters( "woocommerce_after_serializing_{$this->object_type}_for_caching", $data, $object, $id );
+
+		$this->last_cached_data = $data;
+		return $this->get_cache_engine()->cache_object( $this->get_cache_key_prefix() . $id, $data, self::DEFAULT_EXPIRATION === $expiration ? $this->default_expiration : $expiration );
+	}
+
+	/**
+	 * Check if the given expiration time value is valid, throw an exception if not.
+	 *
+	 * @param int $expiration Expiration time to check.
+	 * @return void
+	 * @throws CacheException Expiration time is negative or higher than MAX_EXPIRATION.
+	 */
+	private function verify_expiration_value( int $expiration ): void {
+		if ( self::DEFAULT_EXPIRATION !== $expiration && ( ( $expiration < 1 ) || ( $expiration > self::MAX_EXPIRATION ) ) ) {
+			throw new CacheException( 'Invalid expiration value, must be ObjectCache::DEFAULT_EXPIRATION or a value between 1 and ObjectCache::MAX_EXPIRATION', $this );
+		}
+	}
+
+	/**
+	 * Retrieve a cached object, and if no object is cached with the given id,
+	 * try to get one via get_from_datastore method or by supplying a callback and then cache it.
+	 *
+	 * @param int|string    $id The id of the object to retrieve.
+	 * @param callable|null $get_from_datastore_callback Optional callback to get the object if it's not cached, it must return an object/array or null.
+	 * @param int           $expiration Expiration of the cached data in seconds from the current time, used if an object is retrieved from datastore and cached.
+	 * @return object|array|null Cached object, or null if it's not cached and can't be retrieved from datastore or via callback.
+	 * @throws CacheException Invalid id parameter.
+	 */
+	public function get( $id, callable $get_from_datastore_callback = null, int $expiration = self::DEFAULT_EXPIRATION ) {
```

Comment:
This might just be personal preference, but if most of the time we will end up using concrete versions of `ObjectCache` that implement the `get_from_datastore()` method, would it make more sense to make that the 3rd arg to the function instead of the 2nd?

I'm ok with this as-is, but was thinking about the most common use case, which might be to customize the expiration, not to provide a custom callback, though I might be wrong. What are your thoughts?

Reply:
I think this makes sense. I have reversed the arguments so it's now `get($id, $expiration, $callback)`.

Additionally, I have changed the max expiration allowed from one day to one month.